quil-rs/src/waveform/templates.rs

```diff
@@ -0,0 +1,264 @@
+//! Templates for generating discrete IQ value sequences representing Quil's defined set of waveforms.
+
+use std::f64::consts::PI;
+
+use crate::{imag, real};
+use ndarray::Array;
+use num_complex::Complex64;
+use statrs::function::erf::erf;
+
+use serde::{Deserialize, Serialize};
+
+const J: Complex64 = Complex64::new(0f64, 1f64);
+
+/// Modulate and phase shift waveform IQ data in place.
+pub fn apply_phase_and_detuning(
+    iq_values: &mut [Complex64],
+    phase: f64,
+    detuning: f64,
+    sample_rate: f64,
+) {
+    for (index, value) in iq_values.iter_mut().enumerate() {
+        *value *= (2.0 * PI * J * (detuning * (index as f64) / sample_rate + phase)).exp();
+    }
+}
+
+/// A custom `ceil()` method that includes a machine-epsilon sized region above each integer in the
+/// set of values that are mapped to that integer. In other words:
+/// ceil_eps(x) = n, for all x and integers n s.t. n <= x < (n + n*epsilon)
+///
+/// To handle accumulated floating point errors in sweeps above typical floating point imprecision
+/// we make epsilon 10x larger than floating point epsilon.
+pub(super) fn ceiling_with_epsilon(value: f64) -> f64 {
+    let truncated = value - (value * 10.0 * std::f64::EPSILON);
```

Comment:
Should the adjusted value be the value _plus_ 10*epsilon?

Reply:
I don't think so.